File: core-java/src/main/java/chessgame/domain/model/Board.java
package chessgame.domain.model;

import chessgame.domain.Piece;
import chessgame.domain.Player;
import chessgame.domain.exception.BoardException;
import chessgame.domain.factory.PieceFactory;

import java.util.Arrays;
import java.util.TreeMap;

/**
 * Chessboard.
 *
 * @author Prestes, E. M.
 * @see Piece
 * @see Player
 * @since September 2016
 */
public class Board extends TreeMap<BoardPosition, Piece> {

    private final Player whitePlayer;

    private final Player blackPlayer;

    /**
     * Default constructor
     */
    public Board() {
        super();

        this.whitePlayer = new HumanPlayer();
        this.blackPlayer = new HumanPlayer();
    }

    void init() {
        Arrays.stream(BoardPosition.values())
                .forEach(position -> put(position, null));
    }

    /**
     * Set up a chess game.
     */
    void start() {
        for (BoardPosition position : BoardPosition.values()) {
            switch (position) {
                case A1:
                case H1:
                    put(whitePlayer, PieceFactory.createWhiteRook(this, position));
                    break;
                case B1:
                case G1:
                    put(whitePlayer, PieceFactory.createWhiteKnight(this, position));
                    break;
                case C1:
                case F1:
                    put(whitePlayer, PieceFactory.createWhiteBishop(this, position));
                    break;
                case D1:
                    put(whitePlayer, PieceFactory.createWhiteQueen(this));
                    break;
                case E1:
                    put(whitePlayer, PieceFactory.createWhiteKing(this));
                    break;
                case A2:
                case B2:
                case C2:
                case D2:
                case E2:
                case F2:
                case G2:
                case H2:
                    put(whitePlayer, PieceFactory.createWhitePawn(this, position));
                    break;
                case A7:
                case B7:
                case C7:
                case D7:
                case E7:
                case F7:
                case G7:
                case H7:
                    put(blackPlayer, PieceFactory.createBlackPawn(this, position));
                    break;
                case A8:
                case H8:
                    put(blackPlayer, PieceFactory.createBlackRook(this, position));
                    break;
                case B8:
                case G8:
                    put(blackPlayer, PieceFactory.createBlackKnight(this, position));
                    break;
                case C8:
                case F8:
                    put(blackPlayer, PieceFactory.createBlackBishop(this, position));
                    break;
                case D8:
                    put(blackPlayer, PieceFactory.createBlackQueen(this));
                    break;
                case E8:
                    put(blackPlayer, PieceFactory.createBlackKing(this));
                    break;
                default:
                    put(position, null);
                    break;
            }
        }
    }

    Piece get(String key) {
        return super.get(BoardPosition.valueOf(key));
    }

    /**
     * Put chess piece of a relevant player on the chessboard.
     *
     * @param player Player informed.
     * @param piece  Piece informed.
     * @return Piece
     */
    private Piece put(Player player, Piece piece) {
        player.add(piece);
        return super.put(piece.getPosition(), piece);
    }

    /** {@inheritDoc} */
    @Override
    public Piece put(BoardPosition key, Piece value) {
        if (size() <= BoardPosition.values().length) {
            if (value != null) {
                super.put(value.getPosition(), null);
                ((AbstractPiece) value).setPosition(key);
            }
            return super.put(key, value);
        }

        throw new BoardException("Board created! More positions aren't allowed!");
    }
}

File: core-kotlin/src/main/kotlin/chessgame/domain/model/Board.kt
package chessgame.domain.model

import chessgame.domain.BoardException
import chessgame.domain.Piece
import chessgame.domain.PieceFactory.Companion.createBlackBishop
import chessgame.domain.PieceFactory.Companion.createBlackKing
import chessgame.domain.PieceFactory.Companion.createBlackKnight
import chessgame.domain.PieceFactory.Companion.createBlackPawn
import chessgame.domain.PieceFactory.Companion.createBlackQueen
import chessgame.domain.PieceFactory.Companion.createBlackRook
import chessgame.domain.PieceFactory.Companion.createWhiteBishop
import chessgame.domain.PieceFactory.Companion.createWhiteKing
import chessgame.domain.PieceFactory.Companion.createWhiteKnight
import chessgame.domain.PieceFactory.Companion.createWhitePawn
import chessgame.domain.PieceFactory.Companion.createWhiteQueen
import chessgame.domain.PieceFactory.Companion.createWhiteRook
import chessgame.domain.Player
import chessgame.domain.model.BoardPosition.*
import java.util.*

enum class BoardColor {
    DARK, LIGHT
}

class Board : TreeMap<BoardPosition, Piece?>() {

    private val whitePlayer = HumanPlayer()
    private val blackPlayer = HumanPlayer()

    fun init(): Board {
        values().forEach { position -> put(position, null) }
        return this
    }

    fun start(): Board {
        values().forEach { position ->
            when (position) {
                A1, H1 -> put(whitePlayer, createWhiteRook(this, position))
                B1, G1 -> put(whitePlayer, createWhiteKnight(this, position))
                C1, F1 -> put(whitePlayer, createWhiteBishop(this, position))
                D1 -> put(whitePlayer, createWhiteQueen(this))
                E1 -> put(whitePlayer, createWhiteKing(this))
                A2, B2, C2, D2, E2, F2, G2, H2 -> put(whitePlayer, createWhitePawn(this, position))
                A7, B7, C7, D7, E7, F7, G7, H7 -> put(blackPlayer, createBlackPawn(this, position))
                A8, H8 -> put(blackPlayer, createBlackRook(this, position))
                B8, G8 -> put(blackPlayer, createBlackKnight(this, position))
                C8, F8 -> put(blackPlayer, createBlackBishop(this, position))
                D8 -> put(blackPlayer, createBlackQueen(this))
                E8 -> put(blackPlayer, createBlackKing(this))
                else -> put(position, null)
            }
        }

        return this
    }

    operator fun get(key: String): Piece? {
        return super.get(valueOf(key))
    }

    private fun put(player: Player, piece: Piece?): Piece? {
        piece?.let { player.add(piece) }
        return put(piece!!.position!!, piece)
    }

    override fun put(key: BoardPosition, value: Piece?): Piece? {
        if (size <= BoardPosition.values().size) {
            if (value != null) {
                super.put(value.position!!, null)
                value.position = key
            }
            return super.put(key, value)
        }

        throw BoardException("Board created! More positions aren't allowed!")
    }

}
